File: client/src/game/models/ddraft.ts
interface DDraftCoord {
    x: number;
    y: number;
}

interface DDraftPortal {
    position: DDraftCoord;
    bounds: DDraftCoord[];
    rotation: number; // in radians
    closed: boolean;
    freestanding: boolean;
}

interface DDraftResolution {
    map_origin: DDraftCoord;
    map_size: DDraftCoord;
    pixels_per_grid: number;
}

interface DDraftLight {
    position: DDraftCoord;
    range: number;
    intensity: number;
    color: string;
    shadows: boolean;
}

export interface DDraftData {
    format: number;
    resolution: DDraftResolution;
    line_of_sight: DDraftCoord[][];
    objects_line_of_sight?: DDraftCoord[][];
    portals: DDraftPortal[];
    environment: {
        // these are not supported atm
        baked_lighting: boolean;
        ambient_light: string;
    };
    lights: DDraftLight[];
}

File: server/src/api/socket/asset_manager/ddraft.py
import base64
import json

from typing_extensions import TypedDict


class Coord(TypedDict):
    x: int
    y: int


class DDraftPortal(TypedDict):
    position: Coord
    bounds: list[Coord]
    rotation: int
    closed: bool
    freestanding: bool


class DDraftResolution(TypedDict):
    map_origin: Coord
    map_size: Coord
    pixels_per_grid: int


class DDraftData(TypedDict):
    format: str
    resolution: DDraftResolution
    line_of_sight: list[Coord]
    portals: list[DDraftPortal]
    image: str


def get_ddraft_data(data: bytes):
    ddraft_file: DDraftData = json.loads(data)

    image = base64.b64decode(ddraft_file["image"])

    template = {k: v for k, v in ddraft_file.items() if k != "image"}

    return image, template
